### DISCUSSION TITLE
"when" (joi) in valibot

### DISCUSSION THREAD

**Discussion by abhishekmardiya** [open-circle/valibot#836] [General]
email: Joi.when("_id", {
			is: Joi.exist(),
			then: Joi.optional(),
			otherwise: Joi.string().email().optional(),
		}),
		
		how to achieve similar in valibot

**Comment by fabian-hiller** [comment]
I recommend using `union`, `variant` or `check`, depending on your exact use case. Feel free to try it out in our [playground](https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6BjCAO1XgGUmALAUxDIAjHAC8mAHQkowZKxgAKQgH1gAE0IAaOAG16cSRAoArPk0UBvfQbiq1ALklhgYPgowSOM1gHMFASm0PAFdg9QDA6wMBZGBSRw9nV3dPGG8-QMkYuIjNawBfSIMPI1NzBSsbW3UEiWDWNT5sYFY+NQC8quz4yXrG5tb2orhCvIBdfwZmNg44bn5BACYxXtZgNgU9YolSs0so6ocnFzcPLxaMoLqwoeHowTjapNPU9I6sh9JcguGSkz2Kgc7LU+k0Wm0OgduiCGmDBpCDKN6BMpix2FxeDEAMwrRInBTWP5lfZVYGGMAwdasMgpZ4pc6+d4hG4RO5waHkylsGl45JnNIXJkSbqszojX4STFMADWCgULTAwRg-jEAD44ABCBVKiR2OAAH31mq1rEVMGFn0ik0YaNmUD4qGCpHg4jOyGwfAACsgoKg3PMYkJtJVDo5CNgAOwADj4yAAbABWZAAWgjAAY1BHkwAWKPYITJ+NtHPYbAUIRpxYURbZ7BRrTWDmEYxyPgAAT4AA9kOBSHxJfgG4VUTMIH2JKQIH57Y7ndagA).

```ts
import * as v from 'valibot';

const Schema1 = v.variant('_id', [
  v.object({
    _id: v.pipe(v.string(), v.uuid()),
    email: v.pipe(v.string(), v.email()),
  }),
  v.object({
    _id: v.undefined(),
    email: v.undefined(),
  }),
]);

const Schema2 = v.union([
  v.object({
    _id: v.pipe(v.string(), v.uuid()),
    email: v.pipe(v.string(), v.email()),
  }),
  v.object({
    _id: v.undefined(),
    email: v.undefined(),
  }),
]);

const Schema3 = v.pipe(
  v.object({
    _id: v.optional(v.pipe(v.string(), v.uuid())),
    email: v.optional(v.pipe(v.string(), v.email())),
  }),
  v.check((input) => !input._id || !!input.email)
);
```

Related links:
- https://valibot.dev/guides/unions/
- https://valibot.dev/guides/pipelines/

**Comment by abhishekmardiya** [comment]
Thank you so much! I've been searching for a solution like this for days.

By the way, could you please update this page: [Migrate from Joi](https://valibot.dev/guides/migrate-from-joi/) to make it easier to distinguish the APIs from Joi?

Also, you've built an amazing library—hats off to you!

**Reply by fabian-hiller** [reply]
Thank you very much! Unfortunately, I am not familiar with Joi and need help writing the guide. Just in case you are interested, we could try to figure it out together.

**Reply by abhishekmardiya** [reply]
Yes, I would love to! 
I'm new to Valibot, so once I understand it, I'll definitely help you out.

**Reply by fabian-hiller** [reply]
Thank you! Feel free to reach out once you feel ready!

**Reply by abhishekmardiya** [reply]
```ts
import * as v from 'valibot';

const Schema = v.pipe(
  v.object({
    type: v.string(),
    constructionStatus: v.optional(v.string('require string')),
  }),
  v.check((input) => {
    if (input.type === 'residential') {
      return !!input.constructionStatus;
    }
    return true;
  }, 'Construction Status is required'),
);

const result = v.safeParse(Schema, {
  type: 'residential',
  // constructionStatus: 'Ready to move',
});

console.log(result);
```
RESPONSE:
```ts
[log]: {
  typed: true,
  success: false,
  output: {
    type: "residential"
  },
  issues: [
    {
      kind: "validation",
      type: "check",
      input: {
        type: "residential"
      },
      expected: null,
      received: "Object",
      message: "Construction Status is required",
      requirement: Function,
      path: undefined,
      issues: undefined,
      lang: undefined,
      abortEarly: undefined,
      abortPipeEarly: undefined
    }
  ]
}
```

My goal is to append the errors returned from the check function to the corresponding registration field so that I can identify which specific field the error is associated with. Essentially, I want to display the error as an issue type, like 'constructionStatus'.
e.g.,
```ts
[log]: {
  typed: true,
  success: false,
  output: {
    type: "residential"
  },
  issues: [
    {
      kind: "validation",
      //need to append this type as "constructionStatus"
      type: "check",
      input: {
        type: "residential"
      },
      expected: null,
      received: "Object",
      message: "Construction Status is required",
      requirement: Function,
      path: undefined,
      issues: undefined,
      lang: undefined,
      abortEarly: undefined,
      abortPipeEarly: undefined
    }
  ]
}
```

**Reply by fabian-hiller** [reply]
We have `forward` for this use case: https://valibot.dev/api/forward/

**Reply by abhishekmardiya** [reply]
Thanks.
The forward solution works perfectly for our use case. Appreciate your help!
